File: src/main/python/24444.py
from collections import deque


n,m,r = map(int,input().split())
graph = [[] for _ in range(n+1)]
visited = [0] * (n+1)
for i in range(n):
    u,v = map(int,input().split())
    graph[u].append(v)
    graph[v].append(u)

for i in range(n+1):
    graph[i].sort()


def bfs(graph,R,visited):
    queue = deque([R])
    visited[R] = 1
    cnt = 2
    while queue:
        R = queue.popleft()
        for i in graph[R]:
            if visited[i]==0:
                queue.append(i)
                visited[i] = cnt
                cnt+=1
bfs(graph,r,visited)                
print(*visited[1:],sep="\n")

File: src/main/kotlin/24444.kt
import java.util.LinkedList
import java.util.Queue

fun main(args: Array<String>) {
    val br = System.`in`.bufferedReader()
    val bw = System.out.bufferedWriter()
    val (n,m,r) =  br.readLine().split(' ').map{it.toInt()}
    var visited = Array(n+1){0}
    val graph = Array(n+1){ArrayList<Int>()}
    repeat(m){
        val (u,v) = br.readLine().split(' ').map{it.toInt()}
        graph[u].add(v)
        graph[v].add(u)
    }
    repeat(n+1){
    graph[it].sort()
    }

    fun bfs(nodeNum:Int){
        var que: Queue<Int> = LinkedList()
        que.offer(nodeNum)
        var cnt = 1
        visited[nodeNum] = cnt
        cnt++
        while (que.isNotEmpty()) {
            var temp = que.element()
            que.poll()
            for (i in graph[temp]) {
                if (visited[i]==0) {
                    visited[i] = cnt
                    cnt++
                    que.offer(i)
                }
            }
        }
    }
    bfs(r)
    for(i in 1..n){
        bw.write("${visited[i]}\n")
    }
    bw.close()
}